File: app/src/main/java/br/com/faru/a99challenge/presentation/person/PersonsAdapter.java
package br.com.faru.a99challenge.presentation.person;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.facebook.drawee.view.SimpleDraweeView;

import java.util.ArrayList;
import java.util.List;

import br.com.faru.a99challenge.R;
import br.com.faru.a99challenge.model.Person;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class PersonsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<Person> personList;
    private PersonsFragment.OnPersonClickListener onClickListener;

    public PersonsAdapter(PersonsFragment.OnPersonClickListener onClickListener) {
        this.personList = new ArrayList<>();
        this.onClickListener = onClickListener;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_person, parent, false), onClickListener);
    }

    @Override
    public void onBindViewHolder(final RecyclerView.ViewHolder holder, final int position) {
        ViewHolder viewHolder = (ViewHolder) holder;
        Person person = personList.get(position);
        viewHolder.image.setImageURI(person.getImage());
        viewHolder.name.setText(person.getName());
    }

    public void add(List<Person> personList) {
        this.personList.addAll(personList);
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return personList == null ? 0 : personList.size();
    }

    public Person getItem(int position) {
        return personList.get(position);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.image)
        public SimpleDraweeView image;

        @BindView(R.id.name)
        public TextView name;

        private PersonsFragment.OnPersonClickListener onPersonClickListener;

        public ViewHolder(View itemView, PersonsFragment.OnPersonClickListener onPersonClickListener) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            this.onPersonClickListener = onPersonClickListener;
        }

        @OnClick(R.id.character)
        public void onClick() {
            onPersonClickListener.onClick(getItem(getAdapterPosition()));
        }

    }

}

File: app/src/main/java/br/com/faru/a99challenge/presentation/person/PersonsAdapter.kt
package br.com.faru.a99challenge.presentation.person

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import java.util.ArrayList

import br.com.faru.a99challenge.R
import br.com.faru.a99challenge.model.Person
import com.facebook.drawee.view.SimpleDraweeView

class PersonsAdapter(private val onClick: (p: Person) -> Unit) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val personList: MutableList<Person>

    init {
        this.personList = ArrayList()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context)
                .inflate(R.layout.item_person, parent, false), onClick)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val viewHolder = holder as ViewHolder
        viewHolder.bind(personList[position])
    }

    fun add(personList: List<Person>) {
        this.personList.addAll(personList)
        notifyDataSetChanged()
    }

    override fun getItemCount() = personList.size

    fun getItem(position: Int) = personList[position]

    inner class ViewHolder(itemView: View, private val onClick: (p: Person) -> Unit) : RecyclerView.ViewHolder(itemView) {

        private val image by lazy { itemView.findViewById(R.id.image) as SimpleDraweeView }
        private val name by lazy { itemView.findViewById(R.id.name) as TextView }
        private val character by lazy { itemView.findViewById(R.id.character) as View }

        fun bind(person: Person) {
            image.setImageURI(person.image)
            name.text = person.name
            listener()
        }

        private fun listener() = character.setOnClickListener({ onClick(getItem(adapterPosition)) })
    }

}
